docs/source/more_types.rst

```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+
+Runtime behavior
+----------------
+
+An overloaded function must consist of two or more overload *variants*
+followed by an *implementation*. The variants and the implementations
+must be adjacent in the code: think of them as one indivisible unit.
+
+The variant bodies must all be empty; only the implementation is allowed
+to contain code. This is because at runtime, the variants are completely
+ignored: they're overridden by the final implementation function.
+
+This means that an overloaded function is still an ordinary Python
+function! There is no automatic dispatch handling: you must manually
+handle the different types in the implementation (usually by using
+``if`` statements and ``isinstance`` checks).
+
+If you are adding an overload within a stub file, the implementation
+function should be omitted: stubs do not contain runtime logic.
+
+Type checking calls to overloads
+--------------------------------
+
+When you call an overloaded function, mypy will infer the correct
+return type using the provided variants. A call is never type checked
+against the implementation signature. This is why mypy will report calls
+like ``ip_address(4)`` as being invalid even though it matches the
+implementation signature.
+
+If multiple variants end up matching a call, mypy will, for the
+most part, select the return type corresponding to the first
+matching call. For example, consider the following program:
+
+.. code-block:: python
+
+    from typing import List, Union, overload
+
+    @overload
+    def summarize(data: List[str]) -> str: pass
+
+    @overload
+    def summarize(data: List[int]) -> int: pass
+
+    def summarize(data):
+        # ...snip...
+
+    # What is the type of 'output'? str or int?
+    output = summarize([])
+
+The ``summarize([])`` call matches both variants: an empty list could
+be either a ``List[str]`` or a ``List[int]``. In this case, mypy
+will break the tie by picking the first matching variant: ``output``
+will have an inferred type of ``str``. The implementor is responsible
+for making sure ``summarize`` breaks ties in the same way at runtime.
+
+There are a few exceptions to the "pick the first match" rule.
+For example, if multiple variants match due to an argument
+being of type ``Any``, mypy will make the inferred type also
+be ``Any``.
+
+Mypy will also prohibit you from writing overload variants that are
+inherently unsafely overlapping: for example, writing two variants
+that accept the same arguments but return different types.
+
+Type checking the implementation
+--------------------------------
+
+The body of an implementation is type-checked against the
+type hints provided on the implementation. For example, in the
+``MyList`` example up above, the code in the body is checked with
+``index: Union[int, slice]`` and a return type ``Union[T, Sequence[T]]``.
+If there are no annotations on the implementation, then the body is
+not type checked.
```

Comment:
Maybe clarify that this is different from having all args and return annotated with `Any`. That's standard, but somehow a reminder in this context seems helpful.